Fix inotify related bugs and events specification

Rem01Gaming opened the pull request:
[1] Previously, `IN_CLOSE_WRITE` and `IN_CLOSE_NOWRITE` was represented as `XWATCHER_FILE_REMOVED` which are completely wrong. According to the man page these events indicate that a file was closed (after being opened for writing or reading respectively), not that it was removed.

[2] Fix a race condition on `XWATCHER_FILE_MODIFIED` event where the watcher process immediately read the file when the writing process are not done yet, causing ENOENT error due inode change within write process. Use `IN_CLOSE_WRITE` to represent `XWATCHER_FILE_MODIFIED` instead as it triggered after the file closed for writing.

[3] Use `IN_DELETE_SELF` to represent `XWATCHER_FILE_REMOVED` as it tells when the watched file/dir itself was removed, not the file/dir inside the watched directory. More accurate representation of `XWATCHER_FILE_REMOVED` description in `README.md`.

Rem01Gaming committed:
Fix inotify related bugs and events specification

[1] Previously, `IN_CLOSE_WRITE` and `IN_CLOSE_NOWRITE` was represented as `XWATCHER_FILE_REMOVED` which are completely wrong. According to the man page these events indicate that a file was closed (after being opened for writing or reading respectively), not that it was removed.

[2] Fix a race condition on `XWATCHER_FILE_MODIFIED` event where the watcher process immediately read the file when the writing process are not done yet, causing ENOENT error due inode change within write process. Use `IN_CLOSE_WRITE` to represent `XWATCHER_FILE_MODIFIED` instead as it triggered after the file closed for writing.

[3] Use `IN_DELETE_SELF` to represent `XWATCHER_FILE_REMOVED` as it tells when the watched file/dir itself was removed, not the file/dir inside the watched directory. More accurate representation of `XWATCHER_FILE_REMOVED` description in `README.md`.

Signed-off-by: Rem01Gaming <Rem01_Gaming@proton.me>

nikp123 commented:
Thanks for the fixes.
[1] Seems obvious. When I made that tool I didn't really dig too deep to see if my code's behavior is correct. Thanks for correcting that. Should be merged right away.

[2] Same goes here.

[3] However this one is problematic because I put directory and file watching in the same camp. 
<img width="785" height="147" alt="image" src="https://github.com/user-attachments/assets/2929b100-6854-4782-bf01-8222d29f8625" />

I think the solution would be to seperate those two out in order to provide some granularity.
As in, XWATCHER_DIRECTORY_REMOVED vs XWATCHER_DIRECTORY_CONTENT_REMOVED or something like that. As it seems that the current level of abstraction isn't satisfactory. This is, however, a breaking change and should come with a warning.

nikp123 commented:
Wanted an opinion on if this seems more appropriate.

You can contact me via other means for faster responses.

``@nik123true`` in telegram, because it seems that you have it.

Rem01Gaming commented:
> [3] However this one is problematic because I put directory and file watching in the same camp.


Don't worry both file and directory are treated the same in `IN_DELETE_SELF`

Rem01Gaming committed:
Add more inotify events

- XWATCHER_FILE_MOVED
- XWATCHER_DIRECTORY_CONTENT_REMOVED
- XWATCHER_DIRECTORY_CONTENT_MOVED

Signed-off-by: Rem01Gaming <Rem01_Gaming@proton.me>

nikp123 commented:
Seems fine, but I'm thinking of merging this into a ``staging`` branch because I want some unit tests just so I can get the behavior right before breaking everyone's behavior. Also I need to check how it behaves on Windows as well.

nikp123 commented:
This won't be immediate because I have other things to deal with at the moment, but it should still allow you to use your changes (albeit in a different temporary branch).

nikp123 commented:
I have opened up a new branch just now.

Rem01Gaming committed:
Represent `IN_CREATE` as `XWATCHER_DIRECTORY_CONTENT_CRETAED`

`IN_CREATE` was triggered when a new file or directory was created on the watched directory, not the watched file or directory itself.

Signed-off-by: Rem01Gaming <Rem01_Gaming@proton.me>

nikp123 commented:
Is this fine for merging now?

Rem01Gaming commented:
> Is this fine for merging now?

Yup
